### PR TITLE
feat(stream): add join key in logical pk

### PR DESCRIPTION
**Pull request by st1page**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

As title. Try to solve https://github.com/risingwavelabs/risingwave/issues/5277
## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

### CODE REVIEW

File: src/frontend/src/optimizer/plan_node/logical_join.rs
Comment (fuyufjh) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:123]: Do you mean NOTE?

```suggestion
        // NOTE(st1page): add join keys in the pk_indices a work around before we really have stream
```

File: src/frontend/src/optimizer/plan_node/logical_join.rs
Comment (fuyufjh) [inline src/frontend/src/optimizer/plan_node/logical_join.rs:1149]: Question: So you mean we don't have the actual "stream key" yet 🤔 Why? 

From my understanding the `logical_pk` in PlanNode is the stream key. Am I incorrect?

File: src/frontend/planner_test/tests/testdata/append_only.yaml
Comment (fuyufjh) [inline src/frontend/planner_test/tests/testdata/append_only.yaml:17]: IIUC append-only joins do not have that problem. Maybe should be skipped.

File: src/frontend/src/optimizer/plan_node/logical_join.rs
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/logical_join.rs:1149]: Yes, now we just use `logical_pk` as the `stream_key` and add rewrite the plan tree from bottom to maintain every plan node has `logical_pk` in the locical phase.
But in future we will move the rewriting phase on the stream plan tree to maintain every plan node has `stream_key`. so that the stream_key can refer to `distribution key`, `append_only` and other stream properties

File: src/frontend/planner_test/tests/testdata/append_only.yaml
Comment (fuyufjh) [reply src/frontend/planner_test/tests/testdata/append_only.yaml:17]: Emmm. Why you mark this as resolved?

Comment (st1page) [reply src/frontend/planner_test/tests/testdata/append_only.yaml:17]: err... I think it is explained by another comment
> Yes, now we just use logical_pk as the stream_key and add rewrite the plan tree from bottom to maintain every plan node has logical_pk in the locical phase.
But in future we will move the rewriting phase on the stream plan tree to maintain every plan node has stream_key. so that the stream_key can refer to distribution key, append_only and other stream properties

**Comment by BugenZhao** [comment]
The in-memory e2e failed, which is weird. 😢 Maybe there's something wrong with the implementation of the updated in-memory state store? I'll check it tomorrow.

**Comment by BugenZhao** [comment]
BTW, please only add the `can-merge` label **after** the PR is approved in case there're still some minor issues to be resolved, or some comments from another reviewer are on the way. :)

**Comment by st1page** [comment]
> BTW, please only add the `can-merge` label **after** the PR is approved in case there're still some minor issues to be resolved, or some comments from another reviewer are on the way. :)

ok, got it. I notice that there are lots of `can-merge` labels used as "merge after approve". so maybe we can have a bot to do it?

File: src/frontend/planner_test/tests/testdata/append_only.yaml
Comment (fuyufjh) [reply src/frontend/planner_test/tests/testdata/append_only.yaml:17]: Oh I see

**Comment by BugenZhao** [comment]
> > BTW, please only add the `can-merge` label **after** the PR is approved in case there're still some minor issues to be resolved, or some comments from another reviewer are on the way. :)
> 
> ok, got it. I noticed that there are lots of `can-merge` labels used as "merge after approve". so maybe we can have a bot to remove `can-merge` label for all PR without approve?

I've had an experiment and it does not work. 🤣 https://stackoverflow.com/questions/69493260/get-approval-status-in-github-action

**Comment by fuyufjh** [comment]
```
e2e_test/streaming/time_window_join.slt                      .. [FAILED]
```

E2E test failed. Please take a look @st1page

**Comment by st1page** [comment]
> ```
> e2e_test/streaming/time_window_join.slt                      .. [FAILED]
> ```
> 
> E2E test failed. Please take a look @st1page

Looks running well in hummock state store. 

> The in-memory e2e failed, which is weird. cry Maybe there's something wrong with the implementation of the updated in-memory state store? I'll check it tomorrow.

**Comment by st1page** [comment]
> > ```
> > e2e_test/streaming/time_window_join.slt                      .. [FAILED]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > E2E test failed. Please take a look @st1page
> 
> Looks running well in hummock state store.
> 
> > The in-memory e2e failed, which is weird. cry Maybe there's something wrong with the implementation of the updated in-memory state store? I'll check it tomorrow.

thanks for @BugenZhao's troubleshooting and debugging, he found the vnode is wrong. The memory state store can expose more issues as a really share nothing state store.

**Comment by BugenZhao** [comment]
> > > ```
> > > e2e_test/streaming/time_window_join.slt                      .. [FAILED]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > E2E test failed. Please take a look @st1page
> > 
> > 
> > Looks running well in hummock state store.
> > > The in-memory e2e failed, which is weird. cry Maybe there's something wrong with the implementation of the updated in-memory state store? I'll check it tomorrow.
> 
> 
> thanks for @BugenZhao's troubleshooting and debugging, he found the vnode is wrong. The memory state store can expose more issues as a really share nothing state store.

I've opened a PR for missing vnode check for `Materialize` in #5417.

File: src/frontend/src/optimizer/plan_node/batch_hash_join.rs
Comment (st1page) [inline src/frontend/src/optimizer/plan_node/batch_hash_join.rs:86]: PTAL @fuyufjh

File: src/frontend/planner_test/tests/testdata/join.yaml
Comment (BugenZhao) [inline src/frontend/planner_test/tests/testdata/join.yaml:239]: A lot of `i.x` here. 😄 Do we have an issue for this?

Comment (BugenZhao) [reply src/frontend/planner_test/tests/testdata/join.yaml:239]: Could you please also add a test for a trivial `FullOuter` join? Like 

https://github.com/risingwavelabs/risingwave/blob/7d2651c122bfb11b21f3d4c6eb34ec0625174465/e2e_test/streaming/join.slt#L70-L95

**Review by BugenZhao — COMMENTED** [review]
LGTM

File: src/frontend/planner_test/tests/testdata/join.yaml
Comment (st1page) [reply src/frontend/planner_test/tests/testdata/join.yaml:239]: maybe I will refine the join's planner test later...

File: src/frontend/src/optimizer/plan_node/batch_hash_join.rs
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/batch_hash_join.rs:86]: Make sense. A right join is generated from q15 previously?

Comment (st1page) [reply src/frontend/src/optimizer/plan_node/batch_hash_join.rs:86]: Not. this PR is trying to solve another problem and add the join key into the `stream_key`. after that, some plan has more columns that are not pruned so that the optimizer can derive more distribution and omit exchange. then the issue is exposed.

**Review by BugenZhao — APPROVED** [review]

File: src/frontend/src/optimizer/plan_node/stream_hash_join.rs
Comment (fuyufjh) [inline src/frontend/src/optimizer/plan_node/stream_hash_join.rs:105]: What is this part of the changes for?

**Review by fuyufjh — APPROVED** [review]

File: src/frontend/src/optimizer/plan_node/stream_hash_join.rs
Comment (st1page) [reply src/frontend/src/optimizer/plan_node/stream_hash_join.rs:105]: https://github.com/risingwavelabs/risingwave/pull/5344/files/5bf2ab3e46efcb0fb19bba41d05e552ba7a6b8f5#r973925141

**Pull request merged**
